src/PrestaShopBundle/Form/Admin/Improve/International/Tax/TaxOptionsType.php

```diff
@@ -52,38 +53,89 @@ class TaxOptionsType extends AbstractType
      */
     private $taxRuleGroupChoiceProvider;
 
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
     /**
      * TaxOptionsType constructor.
      *
      * @param bool $ecoTaxEnabled
      * @param FormChoiceProviderInterface $taxAddressTypeChoiceProvider
      * @param FormChoiceProviderInterface $taxRuleGroupChoiceProvider
+     * @param TranslatorInterface $translator
      */
     public function __construct(
         $ecoTaxEnabled,
         FormChoiceProviderInterface $taxAddressTypeChoiceProvider,
-        FormChoiceProviderInterface $taxRuleGroupChoiceProvider
+        FormChoiceProviderInterface $taxRuleGroupChoiceProvider,
+        TranslatorInterface $translator
```

Comment:
BC breaks :thinking:

Reply:
Indeed.

@JevgenijVisockij you need to make this parameter optional `TranslatorInterface $translator = null` to make sure that, if someone has extended this class, this will not break its code.

So
1. make it an optional new argument `,TranslatorInterface $translator = null)`
2. in all the places where you use it, check whether it's null `if ($this->translator === null)`

If however this adds a lot of `if ($this->translator === null)` everywhere, another solution is to deprecate `TaxOptionsType` and create a new class instead. We use the new class but we leave old TaxOptionsType to avoid breaking code that would rely on it.

New class could be `TwigOptimizedTaxOptionsType` for example 😄

Reply:
So I wondered about first option. It's only really being used in one function, buildForm. However if add at the start of the function
```
      if ($this->translator === null) {
            throw new Exception('Fail');
        }
```
It's still a BC break right? If somebody extends TaxOptionType and relies on buildForm now it will throw an exception.
So in that case I should use alternative to $this->translator instead?

Reply:
Yes, you're right.
So I look deeper at the PR, well, without the translator nothing really works.

I see 2 possibilities:
1. use another class, as mentioned above => we let this class here for backward compatibility and use a new one
2. introduce "soft" BC break by using optional arg but add

```
if ($this->translator === null) {
            throw new Exception('This class requires the translator to be injected to work.');
}
```

it's a BC break a little more friendly as it does not make module code crash but it tells developer what it needs

wdty @jolelievre @matthieu-rolland ?

Reply:
Throwing an exception or forcing the parameter is a bit the same to me 🤔. Either way it's a BC break that would cause an error at some point

If we absolutely want to avoid BC break then adding a new class is the only solution, however in this case I would choose the BC break to avoid duplicate code and dead code Especially because we are talking a rather specific form type that has very few changes of being extended

A last alternative would be to introduce a trait or interface to inject the translator via DI automatically, this would avid the BC break from the constructor and if the form type is extended as long as it uses the DI and is not used directly it should be correctly injected as well

Reply:
Can't we use already existing TranslatorAwareTrait for that? 

I don't understand traits that well yet, but you only need to add for it to work, and this won't cause BC break right?

```
        calls:
            - { method: setTranslator, arguments: ['@translator'] }
```

Reply:
If the user needs to change something to make it works, it's a BC break.

Reply:
> Can't we use already existing TranslatorAwareTrait for that?
> 
> I don't understand traits that well yet, but you only need to add for it to work, and this won't cause BC break right?
> 
> ```
>         calls:
>             - { method: setTranslator, arguments: ['@translator'] }
> ```

The idea is the following: let's say someone wants to create a module that provides advanced capabilities for taxes. He sees most of the fields he looks for are available in TaxOptionsType. So he says: "cool ! I will reuse this class" and he creates a `MyModuleTaxType extends TaxOptionsType" and he declares it like this:

```yaml
      class: 'MyModule\MyModuleTaxType'
      arguments:
        - '@=service("prestashop.adapter.legacy.configuration").getBoolean("PS_USE_ECOTAX")'
        - '@=service("prestashop.core.form.choice_provider.tax_address_type_choice_provider")'
        - '@=service("prestashop.core.form.choice_provider.tax_rule_group_choice_provider")'
      tags:
        - { name: form.type }
```

No we modify TaxOptionsType to require the translator for 1.7.8 .

The shop owner, who uses the module, updates prestashop to 1.7.8. He loads the page where MyModuleTaxType is used. The page crashes because the translator needs to be injected.

The module _needs to be fixed_ by adding
```yml
- { method: setTranslator, arguments: ['@translator'] }
```

**If the module developer needs to modify his implementation to "just make it work" it means it was a BC break.**

Reply:
So if we agree, I suggest we follow this:

> I would choose the BC break to avoid duplicate code and dead code Especially because we are talking a rather specific form type that has very few changes of being extended

To do this is pretty simple:
- in the code we do the BC break (maybe identify explicitely with a comment
- we mark this PR as "BC break"